boa_engine/src/object/builtins/jspromise.rs

```diff
@@ -502,44 +471,42 @@ impl JsPromise {
     ///         Ok(JsValue::undefined())
     ///     },
     ///     context,
-    /// )?
+    /// )
     /// .then(
-    ///     Some(
-    ///         FunctionObjectBuilder::new(
-    ///             context.realm(),
-    ///             NativeFunction::from_fn_ptr(|_, args, context| {
-    ///                 args.get_or_undefined(0)
-    ///                     .to_string(context)
-    ///                     .map(JsValue::from)
-    ///             }),
-    ///         )
-    ///         .build(),
-    ///     ),
+    ///     Some(NativeFunction::from_fn_ptr(|_, args, context| {
+    ///         args.get_or_undefined(0)
+    ///             .to_string(context)
+    ///             .map(JsValue::from)
+    ///     })),
     ///     None,
     ///     context,
-    /// )?;
+    /// );
     ///
     /// context.run_jobs();
     ///
     /// assert_eq!(
-    ///     promise.state()?,
+    ///     promise.state(),
     ///     PromiseState::Fulfilled(js_string!("255.255").into())
     /// );
-    ///
-    /// # Ok(())
-    /// # }
     /// ```
     ///
     /// [`Promise.prototype.then`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then
     #[inline]
+    #[allow(clippy::return_self_not_must_use)] // Could just be used to add handlers on an existing promise
     pub fn then(
         &self,
-        on_fulfilled: Option<JsFunction>,
-        on_rejected: Option<JsFunction>,
+        on_fulfilled: Option<NativeFunction>,
+        on_rejected: Option<NativeFunction>,
```

Comment:
Why constrain it to only accept `NativeFunction`? If it's only to reduce the `FunctionObjectBuilder` then we could implement a conveniance method on `NativeFunction`, (maybe `to_js_function(realm)`) that would convert it.

Reply:
Nice idea! I was just cleaning up the API since we mostly use only `NativeFunction`, but having a `to_fn_object` seems a lot more flexible and should also clean up our API :)